aws/rust-runtime/aws-config/src/default_provider.rs

```diff
@@ -178,6 +178,101 @@ pub mod retry_config {
     }
 }
 
+/// Default app name provider chain
+pub mod app_name {
+    use crate::environment::app_name::EnvironmentVariableAppNameProvider;
+    use crate::profile::app_name;
+    use crate::provider_config::ProviderConfig;
+    use aws_types::app_name::AppName;
+
+    /// Default App Name Provider chain
+    ///
+    /// This provider will check the following sources in order:
+    /// 1. [Environment variables](EnvironmentVariableAppNameProvider)
+    /// 2. [Profile file](crate::profile::app_name::ProfileFileAppNameProvider)
+    pub fn default_provider() -> Builder {
+        Builder::default()
+    }
+
+    /// Builder for [DefaultRegionChain]
+    #[derive(Default)]
+    pub struct Builder {
+        env_provider: EnvironmentVariableAppNameProvider,
+        profile_file: app_name::Builder,
+    }
+
+    impl Builder {
+        #[doc(hidden)]
+        /// Configure the default chain
+        ///
+        /// Exposed for overriding the environment when unit-testing providers
+        pub fn configure(mut self, configuration: &ProviderConfig) -> Self {
+            self.env_provider =
+                EnvironmentVariableAppNameProvider::new_with_env(configuration.env());
+            self.profile_file = self.profile_file.configure(configuration);
+            self
+        }
+
+        /// Override the profile name used by this provider
+        pub fn profile_name(mut self, name: &str) -> Self {
+            self.profile_file = self.profile_file.profile_name(name);
+            self
+        }
+
+        /// Build a [DefaultRegionChain]
+        pub async fn app_name(self) -> Option<AppName> {
+            self.env_provider
+                .app_name()
+                .or(self.profile_file.build().app_name().await)
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use crate::provider_config::ProviderConfig;
+        use crate::test_case::no_traffic_connector;
+        use aws_types::os_shim_internal::{Env, Fs};
+
+        #[tokio::test]
+        async fn prefer_env_to_profile() {
+            let fs = Fs::from_slice(&[("test_config", "[default]\nsdk-ua-app-id = wrong")]);
+            let env = Env::from_slice(&[
+                ("AWS_CONFIG_FILE", "test_config"),
+                ("AWS_SDK_UA_APP_ID", "correct"),
+            ]);
+            let app_name = Builder::default()
+                .configure(
+                    &ProviderConfig::empty()
```

Comment:
this should probably be better named but you want `ProviderConfig::no_configuration()` in general in tests

Reply:
(there is no impact here because you override fs and env anyway), but `no_configuration()` ensures that if you _don't_ override those that you won't end up potentially reading the real FS/env during tests causing flakiness and other weird behavior